Switch to git-testament rather than old build.rs

kinnison opened the pull request:
Rather than the previous not-very-informative version string
which was generated by `build.rs` and then not updated unless
`build.rs` was touched, use the new `git-testament` crate which
does a more complete job at compile time via a proc-macro.

Signed-off-by: Daniel Silverstone <dsilvers@digital-scurf.org>

bors commented:
:umbrella: The latest upstream changes (presumably #1711) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #1704) made this pull request unmergeable. Please resolve the merge conflicts.

nrc commented on src/rustup-cli/common.rs:
```diff
@@ -402,11 +404,22 @@ pub fn list_overrides(cfg: &Cfg) -> Result<()> {
+            format!("{} really {}", pkg_version, testament_string)
+        }
+    };
+}
```
I think it would be fine to create the version string from scratch every time, rather than use a lazy static - it's not going to be on a hot path and lazy static is an extra dep and complicates the code

kinnison replied:
The issue is that clap requires a `&str` passed to `.version()` and in `rustup_mode.rs` we return a clap `App` from a function which makes lifetimes a bit hard to deal with.

nrc reviewed (commented):
Seems like a good idea!

kinnison committed:
Switch to git-testament rather than old build.rs

Rather than the previous not-very-informative version string
which was generated by build.rs and then not updated unless
build.rs was touched, use the new `git-testament` crate which
does a more complete job at compile time via a proc-macro.

Signed-off-by: Daniel Silverstone <dsilvers@digital-scurf.org>

nrc reviewed (approved):
Needs rebasing, feel free to merge after that
